veles/secrets/dockerhubpat/detector.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package dockerhubpat contains a Veles Secret type and a Detector for
+// Docker Hub Personal Access Tokens (prefix `dckr_pat_`).
+package dockerhubpat
+
+import (
+	"regexp"
+
+	"github.com/google/osv-scalibr/veles"
+)
+
+// maxTokenLength is the maximum size of a Docker Hub API key.
+const maxTokenLength = 36
+
+// patRe is a regular expression that matches a Docker Hub API key.
+// Docker Hub Personal Access Tokens have the form: `dckr_pat_` followed by 27
+// alphanumeric characters.
+var patRe = regexp.MustCompile(`dckr_pat_[A-Za-z0-9-_-]{27}`)
+
+// dockerLoginCmdRe is a regular expression that matches a Docker Hub API key with an email or username used in switches of docker login command.
+// for example, `docker login -u username -p dckr_pat_{27}`.
+var dockerLoginCmdRe = regexp.MustCompile(`docker\s+login\s+(?:(?:(?:-u|--username)[=\s]+(\S+))|(?:(?:-p|--password)[=\s]+(dckr_pat_[a-zA-Z0-9_-]{27})))\s+(?:(?:(?:-u|--username)[=\s]+(\S+))|(?:(?:-p|--password)[=\s]+(dckr_pat_[a-zA-Z0-9_-]{27})))`)
+
+var _ veles.Detector = NewDetector()
+
+// detector is a Veles Detector.
+type detector struct{}
+
+// NewDetector returns a new Detector that matches
+// Docker Hub Personal Access Tokens.
+func NewDetector() veles.Detector {
+	return &detector{}
+}
+
+func (d *detector) MaxSecretLen() uint32 {
+	return maxTokenLength
+}
+
+func (d *detector) Detect(content []byte) ([]veles.Secret, []int) {
+	var secrets []veles.Secret
+	var offsets []int
+
+	// 1. docker login command username and pat detection
+	dockerLoginCmdMatches := dockerLoginCmdRe.FindAllSubmatchIndex(content, -1)
```

Comment:
Why not  use FindAllSubmatch? Would make the parsing code below a bit simpler and possibly less prone to errors (e.g. it's unclear to me if m[3] could be -1 when m[2] isn't).

The one benefit of SubmatchIndex would be less data copies but the entire secret is at most 36 bytes so I don't think that's a big issue here.